File: src/voyage_geo/config/schema.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ProviderConfig(BaseModel):
    name: str
    enabled: bool = True
    api_key: str | None = None
    model: str | None = None
    base_url: str | None = None
    max_tokens: int | None = None
    temperature: float = 0.7
    rate_limit_rpm: int = 60


class ExecutionConfig(BaseModel):
    concurrency: int = 10
    retries: int = 3
    retry_delay_ms: int = 1000
    timeout_ms: int = 30000
    iterations: int = 1


class QueryConfig(BaseModel):
    count: int = 20
    strategies: list[Literal["keyword", "persona", "competitor", "intent"]] = Field(
        default=["keyword", "persona", "intent"]
    )


class AnalysisConfig(BaseModel):
    analyzers: list[
        Literal[
            "mindshare",
            "mention-rate",
            "sentiment",
            "positioning",
            "rank-position",
            "citation",
            "competitor",
            "narrative",
        ]
    ] = Field(
        default=[
            "mindshare",
            "mention-rate",
            "sentiment",
            "positioning",
            "rank-position",
            "citation",
            "competitor",
            "narrative",
        ]
    )


class ProcessingConfig(BaseModel):
    """Config for the dedicated processing model used in non-execution LLM calls."""

    provider: str = "anthropic"
    model: str = "claude-opus-4-6"
    api_key: str | None = None
    max_tokens: int = 4096
    temperature: float = 0.7


class ReportConfig(BaseModel):
    formats: list[Literal["html", "json", "csv", "markdown"]] = Field(default=["html", "json"])
    include_charts: bool = True
    include_raw_data: bool = False


class VoyageGeoConfig(BaseModel):
    brand: str | None = None
    website: str | None = None
    competitors: list[str] = []
    providers: dict[str, ProviderConfig] = {}
    processing: ProcessingConfig = ProcessingConfig()
    execution: ExecutionConfig = ExecutionConfig()
    queries: QueryConfig = QueryConfig()
    analysis: AnalysisConfig = AnalysisConfig()
    report: ReportConfig = ReportConfig()
    output_dir: str = "./data/runs"
    log_level: Literal["debug", "info", "warning", "error"] = "info"

File: src/config/schema.ts
import { z } from 'zod';

export const providerConfigSchema = z.object({
  name: z.string(),
  enabled: z.boolean().default(true),
  apiKey: z.string().optional(),
  model: z.string().optional(),
  baseURL: z.string().optional(),
  maxTokens: z.number().default(2048),
  temperature: z.number().default(0.7),
  rateLimit: z.object({
    requestsPerMinute: z.number().default(60),
    tokensPerMinute: z.number().default(100000),
  }).default({}),
});

export const executionConfigSchema = z.object({
  concurrency: z.number().default(3),
  retries: z.number().default(3),
  retryDelayMs: z.number().default(1000),
  timeoutMs: z.number().default(30000),
  iterations: z.number().default(1),
});

export const queryConfigSchema = z.object({
  count: z.number().default(20),
  strategies: z.array(z.enum(['keyword', 'persona', 'competitor', 'intent'])).default([
    'keyword',
    'persona',
    'competitor',
    'intent',
  ]),
  categories: z.array(z.enum([
    'recommendation',
    'comparison',
    'best-of',
    'how-to',
    'review',
    'alternative',
    'general',
  ])).default([
    'recommendation',
    'comparison',
    'best-of',
    'how-to',
    'review',
    'alternative',
    'general',
  ]),
});

export const analysisConfigSchema = z.object({
  analyzers: z.array(z.enum([
    'mindshare',
    'mention-rate',
    'sentiment',
    'positioning',
    'rank-position',
    'citation',
    'competitor',
  ])).default([
    'mindshare',
    'mention-rate',
    'sentiment',
    'positioning',
    'rank-position',
    'citation',
    'competitor',
  ]),
});

export const reportConfigSchema = z.object({
  formats: z.array(z.enum(['html', 'json', 'csv', 'markdown'])).default(['html', 'json']),
  includeCharts: z.boolean().default(true),
  includeRawData: z.boolean().default(false),
});

export const voyageGeoConfigSchema = z.object({
  brand: z.string().optional(),
  website: z.string().optional(),
  competitors: z.array(z.string()).default([]),
  providers: z.record(z.string(), providerConfigSchema).default({}),
  execution: executionConfigSchema.default({}),
  queries: queryConfigSchema.default({}),
  analysis: analysisConfigSchema.default({}),
  report: reportConfigSchema.default({}),
  outputDir: z.string().default('./data/runs'),
  logLevel: z.enum(['debug', 'info', 'warn', 'error']).default('info'),
});

export type ProviderConfig = z.infer<typeof providerConfigSchema>;
export type ExecutionConfig = z.infer<typeof executionConfigSchema>;
export type QueryConfig = z.infer<typeof queryConfigSchema>;
export type AnalysisConfig = z.infer<typeof analysisConfigSchema>;
export type ReportConfig = z.infer<typeof reportConfigSchema>;
export type VoyageGeoConfig = z.infer<typeof voyageGeoConfigSchema>;
